### PR TITLE
Response errors should result in a Left of error

### LINKED ISSUE #577: [client] response errors sometimes not made available
Hi,

When using the `updateCheckRun` mutation on github graphql api in one case I get the following response.

```json
{
    "errors": [
      {
        "locations": [
          {
            "line": 3,
            "column": 7
          }
        ],
        "path": [
          "updateCheckRun"
        ],
        "message": "Could not resolve to a node with the global id of 'MDg6Q2hlY2tSdW4xNzk2OTI3MjMx'.",
        "type": "NOT_FOUND"
      }
    ],
    "data": {
      "updateCheckRun": null
    }
  }
```

The relevant graphql for this is:

```graphql
type Mutation {
    updateCheckRun(input: UpdateCheckRunInput!): UpdateCheckRunPayload
}
```

As per this definition ` "updateCheckRun": null` is a perfectly valid response and will json decode correctly. As such, morpheus assumes that everything is fine, however there are in fact errors. I believe the reason `UpdateCheckRunPayload` is nullable is because in the error case it won't be returned.

So I guess the assumption that a successful decode of the response = no errors is incorrect. ie. https://github.com/morpheusgraphql/morpheus-graphql/blob/master/morpheus-graphql-client/src/Data/Morpheus/Client/Fetch.hs#L63 perhaps this success case should also pattern match on an empty error list.

Or you might want 3 responses of, success, success with errors, parse failure.

I'd happy to attempt a PR if you know what the ideal solution is.

### PR DESCRIPTION
**Pull request by AlistairB**
Even when the response `a` can still be successfully parsed.

Closes #577

### CODE REVIEW

File: morpheus-graphql-client/changelog.md [hunk 1]
```diff
@@ -10,6 +10,7 @@
 
 - more likely to rebuild when a file loaded by `defineByDocumentFile` or
   `defineByIntrospectionFile` is changed
+- response errors should always result in a fetch failure [#577](https://github.com/morpheusgraphql/morpheus-graphql/issues/577)
```
Comment (nalchevanidze) [inline morpheus-graphql-client/changelog.md:13]: you can mention yourself: @AlistairB

File: morpheus-graphql-client/test/Case/JSON/Custom/Errors.hs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+      }
+  |]
+
+resolver :: ByteString -> IO ByteString
+resolver = mockApi "JSON/Custom/Errors"
+
+client :: IO (Either String TestQuery)
+client = fetch resolver ()
+
+test :: TestTree
```
Comment (nalchevanidze) [inline morpheus-graphql-client/test/Case/JSON/Custom/Errors.hs:69]: now i think that, if we test errors then in ideal case we test both cases regular errors, and partial response

File: morpheus-graphql-client/src/Data/Morpheus/Client/Fetch.hs [hunk 3]
```diff
@@ -60,7 +60,8 @@ class Fetch a where
     where
       gqlReq = GQLRequest {operationName = Just opName, query = pack strQuery, variables = fixVars (toJSON vars)}
       -------------------------------------------------------------
-      processResponse JSONResponse {responseData = Just x} = Right x
+      processResponse JSONResponse {responseData = Just x, responseErrors = Nothing} = Right x
+      processResponse JSONResponse {responseData = Just x, responseErrors = Just []} = Right x
       processResponse invalidResponse = Left (show invalidResponse)
```
Comment (nalchevanidze) [inline morpheus-graphql-client/src/Data/Morpheus/Client/Fetch.hs:65]: what do you think, maybe we can use JSON "decode"  instead of regular show?

Comment (nalchevanidze) [reply morpheus-graphql-client/src/Data/Morpheus/Client/Fetch.hs:65]: will provide easier understandable error messages.

Comment (AlistairB) [reply morpheus-graphql-client/src/Data/Morpheus/Client/Fetch.hs:65]: So you might do something like:

```haskell
      processResponse JSONResponse {responseData = Just x, responseErrors = Nothing} = Right x
      processResponse JSONResponse {responseData = Just x, responseErrors = Just []} = Right x
      processResponse JSONResponse {responseErrors = err} = Left $ decodeUtf8 $ encode err
```

but you can't do

```haskell
     processResponse invalidResponse = Left $ decodeUtf8 $ encode invalidResponse
```

As `JSONResponse` has an `a` and based on the `Fetch` typeclass we don't know that we have `ToJSON a` constraint. So you would need to either add the `ToJSON a` constraint, which maybe the user doesn't want to provide. Or exclude the `responseData` from the error message.

Based on this issue I think the current implementation is fine. Also, I think unless you prettify the json, a json error message is not too different to a `show` error message. What do you think?

Comment (AlistairB) [reply morpheus-graphql-client/src/Data/Morpheus/Client/Fetch.hs:65]: If you wanted to improve this more, I think the right solution would be to have `fetch` return a `m (Either FetchFailure a)` where (roughly sketched)

```haskell
data FetchFailure = 
  FetchConnectionIssue String
  | ResponseDecodeFailure String
  | ServerGQLErrors (NonEmpty GQLError)
  deriving (Show, ToJSON)
```

This gives the user structured data they can work with if they want, or they can just show / render to JSON. I'd suggest this be done as a separate issue/PR though.

**Comment by AlistairB** [comment]
@nalchevanidze I've done 2/3 and left some comments on the other suggestion. Let know what you think. Thanks.

**Comment by AlistairB** [comment]
@nalchevanidze I had a bit of time so I thought I'd implement something like https://github.com/morpheusgraphql/morpheus-graphql/issues/577#issuecomment-774554821

The missing piece is even though the `GQLError` is returned, it doesn't include arbitrary fields in the error response. ie.

```json
{
  "data": {
    "repository": null
  },
  "errors": [
    {
      "type": "NOT_FOUND",
      "path": [
        "repository"
      ],
      "locations": [
        {
          "line": 7,
          "column": 3
        }
      ],
      "message": "Could not resolve to a Repository with the name 'facebook/reacttt'."
    }
  ]
}
```

I'd like to be able to get the `type` field somehow.

I think to do this I'd like to update `GQLError` to be include arbitrary fields included in the response. However, this type is used through the various packages. I'd probably instead have something like:

```haskell
data FetchGQLError = FetchGQLError {
  fetchError :: GQLError,
  rawError :: HashMap Text Value
}
```

When you aren't busy with your thesis, I'm curious what you think of what I have implemented and this addition. Thanks

**Comment by AlistairB** [comment]
@nalchevanidze I have updated the PR based on https://github.com/morpheusgraphql/morpheus-graphql/issues/577#issuecomment-823654783 . The only difference is I added another case to `FetchError` of `FetchErrorNoResult` where you get no `a` but also there are no errors. Let me know what you think :smile: 

Oh I also updated the stackage to use the latest LTS, I can revert if you want.

**Review by nalchevanidze — APPROVED** [review]

**Comment by nalchevanidze** [comment]
@AlistairB sorry for long delay . there is some merge conflicts when you resolve them feel free to merge it (is approved!).

yeah. would be great if you could also update change-log :)

File: morpheus-graphql-client/test/Case/JSON/Custom/Errors.hs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+test = testCase "test Errors" $ do
+  value <- client
+  assertEqual
+    "test custom Errors"
+    ( Left
+        ( FetchErrorProducedErrors
+            ((GQLError "Failure" [Position {line = 3, column = 7}] Nothing Nothing) :| [])
```
Comment (AlistairB) [inline morpheus-graphql-client/test/Case/JSON/Custom/Errors.hs:84]: The biggest point of uncertainty was here where in master the `GQLError` data constructor is no longer being exported.

I have updated `Data.Morpheus.Types.Internal.AST.Error` to export it again. The other alternative would be to export `Msg (msg)` and `at` then you could use those to build the same `GQLError`.

Can you advise which of these you prefer (or perhaps some other solution)?

**Review by AlistairB — COMMENTED** [review]
Thanks for the approval @nalchevanidze . I have rebased and updated this PR.

There was one area I was uncertain about which I have left a comment. Also the tests for `morpheus-graphql-app` are failing in master, but otherwise I think everything is passing. (And the Deploy Docs build is failing, apparently unrelated to the changes).

File: morpheus-graphql-client/test/Case/JSON/Custom/Errors.hs (hunk 4 above)
Comment (nalchevanidze) [reply morpheus-graphql-client/test/Case/JSON/Custom/Errors.hs]: Hi. please do not expose GQLError.  

perferded approach is to use

```hs
myError :: GQLError
myError = "my error"
```
or 
```hs
myError :: GQLError
myError = "my error" `at` (Position 1 2) 
 ```

Comment (nalchevanidze) [reply morpheus-graphql-client/test/Case/JSON/Custom/Errors.hs]: of course then we should use Pragma "OverloadedStrings" for that.

File: morpheus-graphql-client/test/Case/JSON/Custom/Errors.hs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+client = fetch resolver ()
+
+test :: TestTree
+test = testCase "test Errors" $ do
+  value <- client
+  assertEqual
+    "test custom Errors"
+    ( Left
+        ( FetchErrorProducedErrors
+            (GQLError "Failure" [Position {line = 3, column = 7}] Nothing Nothing :| [])
```
Comment (nalchevanidze) [inline morpheus-graphql-client/test/Case/JSON/Custom/Errors.hs:84]: pleae ue : 
```hs
FetchErrorProducedErrors (("Failure"  `at` Position {line = 3, column = 7}) :| [ ] )
```

Comment (nalchevanidze) [reply morpheus-graphql-client/test/Case/JSON/Custom/Errors.hs:84]: i thinks , this way is also easier to read :)

File: morpheus-graphql-core/src/Data/Morpheus/Types/Error.hs [hunk 6]
```diff
@@ -0,0 +1,16 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+
+module Data.Morpheus.Types.Error
```
Comment (nalchevanidze) [inline morpheus-graphql-core/src/Data/Morpheus/Types/Error.hs:3]: do we really need this ? .  i think having AST is sufficent

File: morpheus-graphql-core/src/Data/Morpheus/Types/Internal/AST/Error.hs [hunk 7]
```diff
@@ -15,7 +15,8 @@ module Data.Morpheus.Types.Internal.AST.Error
     isInternal,
     GQLErrors,
     GQLError
-      ( message,
+      ( GQLError,
```
Comment (nalchevanidze) [inline morpheus-graphql-core/src/Data/Morpheus/Types/Internal/AST/Error.hs:18]: please hide GQLError constructor again, users should use new approach of the error generating.

on the second hand i have flexibility to extend error type how i wish.

File: morpheus-graphql-core/src/Data/Morpheus/Types/Error.hs (hunk 6 above)
Comment (AlistairB) [reply morpheus-graphql-core/src/Data/Morpheus/Types/Error.hs]: Hi, I have removed this module. I think the downside is that this means users of the client library will need to import `Data.Morpheus.Types.Internal.AST` if they want to inspect the `GQLError` type (unless I am missing something). I think inspecting the error returned is pretty normal as you may want to branch based on what error you got.

I think ideally you should not need to import an `Internal` module for what I think is a common use case. Which isn't to say you need `Data.Morpheus.Types.Error`, you could also choose to export `GQLError (message, position)` from other existing modules.

Anyway, I'm not too fussed either way :)

Comment (nalchevanidze) [reply morpheus-graphql-core/src/Data/Morpheus/Types/Error.hs]: maybe later :) for now is ok how  it is.

**Review by nalchevanidze — APPROVED** [review]

**Pull request merged**